pyra/hardware.py

```diff
@@ -0,0 +1,422 @@
+"""
+..
+   hardware.py
+
+Functions related to the dashboard viewer.
+"""
+# standard imports
+import json
+
+# lib imports
+import GPUtil
+import plotly
+import psutil
+
+# local imports
+from pyra import definitions
+from pyra import helpers
+from pyra import locales
+from pyra import logger
+
+_ = locales.get_text()
+chart_translations = dict(
+    cpu=dict(
+        bare=_('cpu'),
+        usage=_('cpu usage')
+    ),
+    gpu=dict(
+        bare=_('gpu'),
+        usage=_('gpu usage')
+    ),
+    memory=dict(
+        bare=_('memory'),
+        usage=_('memory usage')
+    ),
+    network=dict(
+        bare=_('network'),
+        usage=_('network usage')
+    ),
+    general=dict(
+        received=_('received'),
+        sent=_('sent'),
+        system=_('system')
+    )
+)
+
+log = logger.get_logger(__name__)
+
+initialized = False
+network_recv_last = 0
+network_sent_last = 0
+proc = psutil.Process()  # the main retroarcher process
+proc_id = proc.pid
+processes = [proc]
+
+nvidia_gpus = GPUtil.getGPUs()
+
+try:
+    import pyamdgpuinfo
+except ModuleNotFoundError:
+    amd_gpus = 0
+else:
+    amd_gpus = pyamdgpuinfo.detect_gpus()  # this will be an integer representing the count of amd gpus
+
+dash_stats = dict(
+    timestamp=[],
+    relative_time=[],
+    cpu_system=[],
+    memory_system=[],
+    network_sent=[],
+    network_received=[]
+)
+
+history_length = 120
+
+
+def update_cpu() -> float:
+    """
+    Update dashboard stats for system CPU usage.
+
+    This will append a new value to the ``dash_stats['cpu_system']`` list.
+
+    Returns
+    -------
+    float
+        The current system cpu percentage utilized.
+
+    Examples
+    --------
+    >>> update_cpu()
+    """
+    cpu_percent = psutil.cpu_percent(interval=None, percpu=False)
+
+    if initialized:
+        dash_stats['cpu_system'].append(cpu_percent)
+
+    return cpu_percent
+
+
+def update_gpu():
+    """
+    Update dashboard stats for system GPU usage.
+
+    This will create new keys for the ``dash_stats`` dictionary if required, and then append a new value to the
+    appropriate list.
+
+    Nvidia GPUs are fully supported. AMD GPUs are currently only supported on Linux.
+
+    Examples
+    --------
+    >>> update_gpu()
+    """
+    global nvidia_gpus
+    nvidia_gpus = GPUtil.getGPUs()  # need to get the GPUs again otherwise the load does not update
+
+    for gpu in nvidia_gpus:
+        name = f'{gpu.name}-{gpu.id}'
+
+        if initialized:
+            try:
+                dash_stats[f'gpu_{name}']
+            except KeyError:
+                dash_stats[f'gpu_{name}'] = []
+            finally:
+                gpu_load = 100 if gpu.load > 1 else gpu.load * 100  # convert decimal to percentage, max of 100
+                dash_stats[f'gpu_{name}'].append(gpu_load)
+
+    # this is untested
```

Comment:
This section below could be re-written:
``for amd_index in range(amd_gpus):
    gpu = pyamdgpuinfo.get_gpu(amd_counter)
    name = f'gpu_{gpu.name}-{gpu.gpu_id}'
    if initialized:
        if not dash_stats[name]:
            dash_stats[name] = []
        gpu_load = min(100, gpu.query_load())  # max of 100
        dash_stats[name].append(gpu_load)``